src/python/pants/engine/scheduler.py

```diff
@@ -41,70 +41,39 @@ class ExecutionRequest(datatype('ExecutionRequest', ['roots'])):
-    :param native: An instance of engine.subsystem.native.Native.
-    :param graph_lock: A re-entrant lock to use for guarding access to the internal product Graph
-                       instance. Defaults to creating a new threading.RLock().
-    """
-    self._products_by_goal = goals
-    self._project_tree = project_tree
-    self._native = native
-    self._product_graph_lock = graph_lock or threading.RLock()
-    self._run_count = 0
-
+class WrappedNativeScheduler(object):
+  def __init__(self, native, rule_index, root_subject_types):
     # TODO: The only (?) case where we use inheritance rather than exact type unions.
     has_products_constraint = SubclassesOf(HasProducts)
 
-    # Create the ExternContext, and the native Scheduler.
+    self._native = native
     self._scheduler = native.new_scheduler(has_products_constraint,
                                            constraint_for(Address),
                                            constraint_for(Variants))
-    self._execution_request = None
-
-    # Validate and register all provided and intrinsic tasks.
-    # TODO: This bounding of input Subject types allows for closed-world validation, but is not
-    # strictly necessary for execution. We might eventually be able to remove it by only executing
-    # validation below the execution roots (and thus not considering paths that aren't in use).
-
-    root_subject_types = {
-      Address,
-      BuildFileAddress,
-      AscendantAddresses,
-      DescendantAddresses,
-      PathGlobs,
-      SiblingAddresses,
-      SingleAddress,
-    }
-    intrinsics = create_fs_intrinsics(project_tree) + create_snapshot_intrinsics(project_tree)
-    singletons = create_snapshot_singletons(project_tree)
-    rule_index = RuleIndex.create(tasks, intrinsics, singletons)
-
     self._register_tasks(rule_index.tasks)
     self._register_intrinsics(rule_index.intrinsics)
     self._register_singletons(rule_index.singletons)
+    self.root_subject_types = root_subject_types
 
-    self._validate_ruleset(root_subject_types)
+  def graph_trace(self):
+    with temporary_file_path() as path:
+      self._native.lib.graph_trace(self._scheduler, bytes(path))
+      with open(path) as fd:
+        for line in fd.readlines():
+          yield line.rstrip()
 
-    RulesetValidator(rule_index, goals, root_subject_types).validate()
+  def validate(self):
+    self.assert_ruleset_valid()
 
-  def _validate_ruleset(self, root_subject_types):
-    listed = list(TypeId(self._to_id(t)) for t in root_subject_types)
+  def assert_ruleset_valid(self):
+    listed = list(TypeId(self._to_id(t)) for t in self.root_subject_types)
 
-    self._native.lib.validator_run(self._scheduler, listed, len(listed))
+    raw_value = self._native.lib.validator_run(self._scheduler, listed, len(listed))
+    value = self._from_value(raw_value)
+    if isinstance(value, Exception):
+      raise ValueError(str(value))
+    else:
+      pass # validation success
```

Comment:
Makes sense. Writing more in Rust has brought out my "be explicit around branches" habits.